### ISSUE TITLE
Feature Request: Support for depthwise convolution by groups

### ISSUE THREAD

**Issue by fculinovic** [tensorflow/tensorflow#3332]
As much as I have managed to follow the API section of tensorflow the depthwise_conv2d doesn't fulfill my thoughts on what I would like to do with the input/filters.

```
shape = (2,5,5,48,128)
initializer = tf.truncated_normal_initializer(stddev=1e-2)
kernel = tf.get_variable(name='weights', shape=shape, initializer=initializer)
```

Considering the input is of shape (batch_size, 55, 55, 96),
the function definition would be:

```
def depthwise_group_conv2d(input, filter, groups, strides, padding, name)
```

which would be able to do

```
conv = tf.nn.depthwise_group_conv2d(input, kernel, [48,48], strides, padding, name)
```

that would split the input depthwise into depths from given 'groups' parameter and perform the convolution with shared parameters (i.e. depths [:48] take weights[0] and depths[48:] take weights[1]) inside a group. Then the outputs of convolutions by groups would be concatenated depthwise.

This would make it easier to define groups such as one used in AlexNet/CaffeNet architectures.

Hopefully I have missed a certain feature already existing for making this easier.

Best regards,
Filip

**Comment by jmchen-g** [comment]
I am not sure how to read the numbers in your code. Are they matching? 
Are you trying to combine two depthwise conv to one essentially?

**Comment by fculinovic** [comment]
What I would basically like to do is use convolution on an input of size (batch_size, 55, 55, 96) such that 1 kernel is shared for the whole input UP TO depth 48, and a different kernel from depth 48. Each convolution should therefore give a result of depth 128 because of the defined weights. Then when the results are concatenated depthwise they would give the depth of 256 such as defined in [(http://vision.stanford.edu/teaching/cs231b_spring1415/slides/alexnet_tugce_kyunghee.pdf)].

Currently the implementation of depthwise_conv2d in tensorflow allows for only applying a different filter of depth 1 to each input channel, while I would like to apply a filter of depth 48 to 48 input channels, and then again a different filter of depth 48 for the next 48 input channels.

Such convolution (splitting the input into 2 groups by depth) is used in AlexNet/CaffeNet and it would be great if such a function would be implemented in TensorFlow.

**Comment by jmchen-g** [comment]
OK. Depthwise convolution is actually a different type of conv. You can find "depthwise_conv2d_native".

For this, it might be better to just have two separate convs like what inception did to avoid all the slicing operations.

**Comment by fculinovic** [comment]
Sorry for misunderstanding the depthwise convolution, I thought the term would just generally describe using different filters for different depths (and therefore include groups).
The exact reason to implement a function such as this is to avoid explicit slicing of the kernels, biases and inputs. An additional reason why I thought it might be useful is because Caffe had "groups" as a possible parameter in the convolution layer [http://caffe.berkeleyvision.org/tutorial/layers.html](url). I would like to emulate this behavior.
The exact reason to implement a function such as this is to avoid explicit slicing of the kernels, biases and inputs. An additional reason why I thought it might be useful is because Caffe had "groups" as a possible parameter in the convolution layer [http://caffe.berkeleyvision.org/tutorial/layers.html](url). I thought it might be nice to have tensorflow support for this behavior.

**Comment by jmchen-g** [comment]
I think it is actually logically cleaner to have two separate convs just like what's in inception. We might need a strong use case to support another interface for the very popular op so please send this over when it is the case. Thanks.

**Comment by bowang** [comment]
I implemented this feature in the following PR.
https://github.com/tensorflow/tensorflow/pull/10482

**Comment by Timeroot** [comment]
Nvidia has actually started a campaign advertising hardware specifically based on its ability to do groupwise convolutions. If that doesn't mean that these are mainstream enough to desparately warrant inclusion as an op, I don't know what would.

(I'm sorry if this sounds like whining, my desire is only to point out that it is has gone 'mainstream'! So that when the devs' busy schedules have room, they might prioritize this a bit more. :) Sadly I don't know enough about the internals to contribute, sorry :( )

**Comment by ondrejba** [comment]
The state-of-the-art in image classification (ResNetXt and AttentionNetXt) use grouped convolutions. I think that's a reason enough for them to be added to Tensorflow core.

**Comment by idofr** [comment]
There are 3 threads on this topic, all are closed without any proper conclusion. Is there any intention there?

**Comment by jnjaby** [comment]
Group  convolution is implemented in both Caffe and PyTorch. As now it slows down the inference stage when I reproduce them with `tf.split` and `tf.concat`, I think it's reasonable enough to be added as TF kernel.

**Comment by cancan101** [comment]
Any chance of re-opening this issue? As pointed out, a number of SoTA models use these group convolutions. Further [per this comment](https://github.com/tensorflow/tensorflow/issues/12052#issuecomment-320465264), cuDNN 7 now supports group convolutions as a primitive. It seems like exposing a primitive op that handles these convolutions makes sense.

**Comment by jmchen-g** [comment]
I am no longer in this team and not sure if anyone else is looking at this so just reopened it. :)

**Comment by aselle** [comment]
@martinwicke, @shlens, @fchollet, is there any interest in defining an api for this?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by shlens** [comment]
I suspect that the API does not need to be expanded. Instead, I would propose the following:
- expose [`depthwise_conv2d`](https://github.com/tensorflow/tensorflow/blob/r1.4/tensorflow/python/ops/nn_impl.py#L327) in`nn_impl.py`
- offer an optional integer argument named `group` that defaults to 1.

This would conceptually group together all of the depthwise convolution operations by offering a single function. In addition, this would be a backward-compatible change but also support the type of grouping proposed in the other work. (Under the hood, `depthwise_conv2d` would have a switch statement that would call the appropriate kernel depending on the value of `group`.)

**Comment by x10000year** [comment]
Really want a fast group convolution implementation for ResNeXt!

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activityand the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by Remi-Coulom** [comment]
I also want group convolution in tensorflow. I found that tweet about efficiently implementing ResNeXt:
https://twitter.com/D_Berthelot_ML/status/907007235577880576
"Thanks @VahidK for the solution, using depthwise convolution with a reshape+sum does the trick. It's 4-5x faster now."
That may not be the ideal solution (maybe not as good as calling cudnn's group convolution primitive), but it may help.

**Comment by martinwicke** [comment]
@josh11b @fchollet @asimshankar @sherrym Opinions on grouped convolutions? Maybe especially @sherrym?

**Comment by sherrym** [comment]
I like Jon's suggestion on expanding the depthwise_conv2d API to support the group functionality. Is that not sufficient?

**Comment by josh11b** [comment]
API Review Committee is supportive of a new group argument added to the existing depthwise_conv2D API to address this use case. If someone wants to revive #10482, it looks like it changes the API without modifying the underlying kernels (which would be another step needed for full performance).

**Comment by cancan101** [comment]
Perhaps that issue can be re-opened (CC @jhseu)

**Comment by bowang** [comment]
@josh11b I am happy to revive #10482. Please let me know any improvement you'd like to have on that PR.

**Comment by fchollet** [comment]
The API change is fine, but merging a PR that simply iteratively applies `Conv2D` would result in extremely slow runtime, which would then be misleading. One of the main motivations for using convolution groups is that they should be faster than regular convolutions (separable convolution being the fastest), and in this case we would actually be much slower than regular convolution, by a large factor. 

In my opinion this change only makes sense if we release a kernel that implements it with decent performance.

**Comment by cancan101** [comment]
Perhaps using https://github.com/tensorflow/tensorflow/issues/3332#issuecomment-358818076 to minimize speed hit relative to native kernel?

**Comment by fchollet** [comment]
@cancan101 yes, that would probably be okay. But would it ever be replaced with a fast native kernel?

**Comment by bowang** [comment]
@fchollet I agree. Given CuDNN 7 is being fully adopted by TensorFlow now, we should use its native support for group-based convolution rather than doing things in Python.

**Comment by martinwicke** [comment]
That would be fantastic. If you were to revive #10482, an enhancement to the kernel to actually use the cuDNN primitive would be the way to go.

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.

**Comment by bowang** [comment]
I looked a bit into the usage of CuDNN in TensorFlow, especially `cudnnConvolutionForward`. It's a piece of code of modest size and contains complicated logics for algorithm selection. Since not all forward convolution algorithms support group-based convolution, I think it'll be faster for someone who is familiar with that part of code to add this functionality.

Also, to support an API change in C++ level, we also need to implement the CPU-only version. I'm afraid I don't have enough time for this in the near term.

I'm happy to flesh out a pure API change patch in Python (i.e. without the actual C++ impl) based on #10482 if that seems useful. Otherwise, please feel free to use any code in that #10482.

**Comment by tatianashp** [comment]
@zheng-xq Are you interested in adding fast implementation of group convolution that maps to CuDNN 7?

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by yzhwang** [comment]
Reassigned to @chsigg as he has been working on this.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chsigg: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by Remi-Coulom** [comment]
Any news? CUDNN 7.1.1 has even better support for convolution groups:
http://docs.nvidia.com/deeplearning/sdk/cudnn-release-notes/rel_711.html#rel_711
"All algorithms for convolutions now support groupCount > 1"
This should make things easier.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chsigg: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chsigg: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by TylerBalsam** [comment]
We've been getting a lot of mileage out of the MobileNet class of architectures, especially for edge inference. Any updates on this so we can really use separable convs near their theoretical speed?

**Comment by wendazhou** [comment]
@TylerBalsam I have implemented the CUDNN grouped conv API for DepthwiseConv in https://github.com/tensorflow/tensorflow/pull/17961, you can give it a try if you are willing to compile yourself. I get a significant speed-up for MobileNet (my main application), especially in 16-bit training.

**Comment by TylerBalsam** [comment]
Many thanks!

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chsigg: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chsigg: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chsigg: It has been 49 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by lubomir1** [comment]
Please add support for group convolutions! Caffe and PyTorch have had an efficient implementation for nearly two years, and the current state of the art models leverage groups. Even though one can use depthwise_conv2d_native followed by a sum, this is a very slow and memory-demanding way to implement it, even with the latest CuDNN optimization.

I tried a group convolution with 256 channels and 32 groups, 3x3 kernel and in PyTorch it ran x2 the speed of full convolution (groups==1), whereas if I do it in tf with depthwise_conv2d_native it is 5 times slower relative to full convolution -- so PyTorch is 10 times faster in this case, which is very common in ResNeXt style architectures. I am using the latest TF build and have enabled the CuDNN support for it with a kernel label map.

**Comment by thefiddler** [comment]
Same here, this feature is a blocker for training & deploying many relevant state of the art models on tensorflow. This should seriously be bumped in priority.

**Comment by chsigg** [comment]
Sorry for the large delay on this issue. I will start working on implementing support for grouped convolutions  very soon.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chsigg: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chsigg: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by Timen** [comment]
I implemented a basic version of grouped convolutions using normal convolutions and splitting the input tensor and weights tensor. The implementation is kinda compatible with tensorflow slim. `import tensorflow as tf
from tensorflow.contrib.layers.python.layers import utils
import collections

slim = tf.contrib.slim


def grouped_convolution2D(inputs, filters, padding, num_groups,
                          strides=None,
                          dilation_rate=None):
    """
    Performs a grouped convolution by applying a normal convolution to each of the seperate groups
    :param inputs:
        Input of the shape [<batch_size>,H,W,inC]
    :param filters:
        [H,W,inC/num_groups,outC]
    :param padding:
        What padding to use
    :param num_groups:
        Number of seperate groups
    :param strides:
        Stride
    :param dilation_rate:
        Dilation rate
    :return:
        Output of shape [<batch_size>,H/stride,W/stride,outC]
    """
    # Split input and outputs along their last dimension
    input_list = tf.split(inputs, num_groups, axis=-1)
    filter_list = tf.split(filters, num_groups, axis=-1)
    output_list = []

    # Perform a normal convolution on each split of the input and filters
    for conv_idx, (input_tensor, filter_tensor) in enumerate(zip(input_list, filter_list)):
        output_list.append(tf.nn.convolution(
            input_tensor,
            filter_tensor,
            padding,
            strides=strides,
            dilation_rate=dilation_rate,
            name="grouped_convolution" + "_{}".format(conv_idx)
        ))
    # Concatenate ouptputs along there last dimentsion
    outputs = tf.concat(output_list, axis=-1)

    return outputs


@slim.add_arg_scope
def grouped_convolution(inputs,
                        num_outputs,
                        kernel_size,
                        groups,
                        stride=1,
                        padding='SAME',
                        rate=1,
                        activation_fn=tf.nn.relu,
                        normalizer_fn=None,
                        normalizer_params=None,
                        weights_initializer=tf.contrib.layers.xavier_initializer(),
                        weights_regularizer=None,
                        biases_initializer=tf.initializers.zeros(),
                        biases_regularizer=None,
                        reuse=None,
                        trainable=True,
                        scope=None,
                        outputs_collections=None):
    """Adds an 2-D grouped convolution followed by an optional batch_norm layer.
      `convolution` creates a variable called `weights`, representing the
      convolutional kernel, that is convolved (actually cross-correlated) with the
      `inputs` to produce a `Tensor` of activations. If a `normalizer_fn` is
      provided (such as `batch_norm`), it is then applied. Otherwise, if
      `normalizer_fn` is None and a `biases_initializer` is provided then a `biases`
      variable would be created and added the activations. Finally, if
      `activation_fn` is not `None`, it is applied to the activations as well.
      Performs atrous convolution with input stride/dilation rate equal to `rate`
      if a value > 1 for any dimension of `rate` is specified.  In this case
      `stride` values != 1 are not supported.
      Args:
        inputs: A Tensor of rank N+2 of shape
          `[batch_size] + input_spatial_shape + [in_channels]` if data_format does
          not start with "NC" (default), or
          `[batch_size, in_channels] + input_spatial_shape` if data_format starts
          with "NC".
        num_outputs: Integer, the number of output filters.
        kernel_size: A sequence of N positive integers specifying the spatial
          dimensions of the filters.  Can be a single integer to specify the same
          value for all spatial dimensions.
        groups: Number of groups to split the input up in before applying convolutions to the
            seperate groups. If groups==1 return normal slim.conv2d.
        stride: A sequence of N positive integers specifying the stride at which to
          compute output.  Can be a single integer to specify the same value for all
          spatial dimensions.  Specifying any `stride` value != 1 is incompatible
          with specifying any `rate` value != 1.
        padding: One of `"VALID"` or `"SAME"`.
        rate: A sequence of N positive integers specifying the dilation rate to use
          for atrous convolution.  Can be a single integer to specify the same
          value for all spatial dimensions.  Specifying any `rate` value != 1 is
          incompatible with specifying any `stride` value != 1.
        activation_fn: Activation function. The default value is a ReLU function.
          Explicitly set it to None to skip it and maintain a linear activation.
        normalizer_fn: Normalization function to use instead of `biases`. If
          `normalizer_fn` is provided then `biases_initializer` and
          `biases_regularizer` are ignored and `biases` are not created nor added.
          default set to None for no normalizer function
        normalizer_params: Normalization function parameters.
        weights_initializer: An initializer for the weights.
        weights_regularizer: Optional regularizer for the weights.
        biases_initializer: An initializer for the biases. If None skip biases.
        biases_regularizer: Optional regularizer for the biases.
        reuse: Whether or not the layer and its variables should be reused. To be
          able to reuse the layer scope must be given.
        outputs_collections: Collection to add the outputs.
        trainable: If `True` also add variables to the graph collection
          `GraphKeys.TRAINABLE_VARIABLES` (see tf.Variable).
        scope: Optional scope for `variable_scope`.
      Returns:
        A tensor representing the output of the operation.
      Raises:
        ValueError: If `data_format` is invalid.
        ValueError: Both 'rate' and `stride` are not uniformly 1.
        ValueError: If 'groups'<1.
      """
    # if no group size specified or less than/equal to zero return a normal convolution
    if groups == 1:
        return slim.conv2d(inputs, num_outputs, kernel_size, stride=stride, padding=padding,
                           activation_fn=activation_fn,
                           normalizer_fn=normalizer_fn,
                           normalizer_params=normalizer_params,
                           weights_initializer=weights_initializer,
                           weights_regularizer=weights_regularizer,
                           biases_initializer=biases_initializer,
                           biases_regularizer=biases_regularizer,
                           reuse=reuse,
                           trainable=trainable,
                           scope=scope)
    if groups < 1:
        raise ValueError("Specify a number of groups greater than zero, groups given is {}".format(groups))

    input_channels = inputs.get_shape().as_list()[-1]

    # check if the number of groups and corresponding group_size is an integer division of the input and output channels
    lowest_channels = min(input_channels, num_outputs)
    assert lowest_channels % groups == 0, "the remainder of min(input_channels,output_channels)/groups should be zero"
    assert max(input_channels,
               num_outputs) % groups == 0, "the remainder of max(input_channels,output_channels)/groups=({}) " \
                                               "should be zero".format(
        groups)

    with tf.variable_scope(scope, 'Group_Conv', [inputs], reuse=reuse) as sc:
        # define weight shape
        if isinstance(kernel_size, collections.Iterable):
            weights_shape = list(kernel_size) + [input_channels/groups] + [num_outputs]
        else:
            weights_shape = [kernel_size, kernel_size, input_channels/groups, num_outputs]

        # create weights variable
        weights = slim.variable('weights',
                                shape=weights_shape,
                                initializer=weights_initializer,
                                regularizer=weights_regularizer,
                                trainable=trainable)
        strides = [stride, stride]
        dilation_rate = [rate, rate]
        # perform grouped convolution
        outputs = grouped_convolution2D(inputs, weights, padding, groups,
                                        strides=strides,
                                        dilation_rate=dilation_rate)
        if biases_initializer is not None:
            biases = slim.variable('biases',
                                   shape=[num_outputs],
                                   initializer=biases_initializer,
                                   regularizer=biases_regularizer,
                                   trainable=trainable)
            outputs = tf.nn.bias_add(outputs, biases)
        if normalizer_fn is not None:
            normalizer_params = normalizer_params or {}
            outputs = normalizer_fn(outputs, **normalizer_params)

        if activation_fn is not None:
            outputs = activation_fn(outputs)
        return utils.collect_named_outputs(outputs_collections, sc.name, outputs)`

**Comment by ondrejba** [comment]
I tried something similar but it's too slow.

**Comment by michaelklachko** [comment]
@Timen it seems like you did a simple loop over the inputs. What makes you think your convolutions will run in parallel?

**Comment by Timen** [comment]
@michaelklachko The loop is only there to define the convolutional nodes in the graph, during training the loop will not get executed. These convolutions all consume an independent input (after the split) and produce independent output (with respect to the other convolutions) these outputs are then gathered and combined with a concat. And according to this stackoverflow post https://stackoverflow.com/questions/41233635/meaning-of-inter-op-parallelism-threads-and-intra-op-parallelism-threads tensorflow does attempt to execute nodes that don't have a path linking them in parallel.

**Comment by lubomir1** [comment]
In theory yes, but in practice implementing group convolution in this way is very inefficient. This is why NVidia explicitly provided acceleration of group convolutions in the latest CuDNN. There is a [group option](https://blog.yani.io/cudnn-filter-groups) for nearly a year now.
 All the Tensorflow developers need to do is propagate it to the parameters of conv2d.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chsigg: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chsigg: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chsigg: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chsigg: It has been 60 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by HongyangGao** [comment]
Currently, mobilenetv2 is using this code for split_conv. Hope there will be an op to support group convs.

**Comment by HouseOfFinwe** [comment]
How is progress on this coming along?

**Comment by albertz** [comment]
Note that GPU ops do *not* run in parallel, even with `inter_op_parallelism_threads > 1`, because currently TF uses only a single CUDA compute stream.
See [here](https://stackoverflow.com/questions/51494911/will-tensorflow-matmul-run-in-parallel-on-gpu-or-any-gpu-ops) or [here](https://stackoverflow.com/questions/36098947/how-does-tensorflow-support-cuda-streams) for reference.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chsigg: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by vcarpani** [comment]
Is someone working on this? I would like to work on it, but I do not want to duplicate the work of someone else

**Comment by ondrejba** [comment]
@vcarpani This issue has been dead for a long time, so I think you're fine.
Thanks!

**Comment by tensorflowbutler** [comment]
Nagging Assignee @chsigg: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by rockclimber112358** [comment]
Any update on this?

**Comment by Robird** [comment]
Also need group conv !

**Comment by ppwwyyxx** [comment]
Oh, so this issue has been open for 2.5 years.

The only thing needed to enable group convolution in tensorflow is to remove two lines of code:
```diff
diff --git i/tensorflow/core/framework/common_shape_fns.cc w/tensorflow/core/framework/common_shape_fns.cc
index 5c974a76ac..aa4a663c5f 100644
--- i/tensorflow/core/framework/common_shape_fns.cc
+++ w/tensorflow/core/framework/common_shape_fns.cc
@@ -495,8 +495,6 @@ Status Conv2DShapeImpl(shape_inference::InferenceContext* c,
   // Check that the input tensor and the filter tensor agree on the input
   // channel count.
   DimensionHandle unused;
-  TF_RETURN_IF_ERROR(
-      c->Merge(input_depth_dim, filter_input_depth_dim, &unused));
 
   Padding padding;
   TF_RETURN_IF_ERROR(c->GetAttr("padding", &padding));
diff --git i/tensorflow/core/grappler/costs/op_level_cost_estimator.cc w/tensorflow/core/grappler/costs/op_level_cost_estimator.cc
index 1e2e160955..745039bf27 100644
--- i/tensorflow/core/grappler/costs/op_level_cost_estimator.cc
+++ w/tensorflow/core/grappler/costs/op_level_cost_estimator.cc
@@ -616,7 +616,6 @@ OpLevelCostEstimator::ConvolutionDimensionsFromInputs(
   // Only check equality when both sizes are known (in other words, when
   // neither is set to a minimum dimension size of 1).
   if (iz != 1 && filter_shape.dim(in_channel_index).size() != 1) {
-    CHECK_EQ(iz, filter_shape.dim(in_channel_index).size());
   } else {
     iz = std::max<int64>(iz, filter_shape.dim(in_channel_index).size());
   }
```

It'll take a lot more to make a PR, with tests, shape checks, flops calculation, etc, and probably handle some more edge cases. I'm not sure I'll have time for that. But in case someone needs it and is OK with building TF from source, here you go.
On a ResNeXt50-32x4d trained on ImageNet, I observed a 13% overall speed improvement vs a loop-based implementation.

**Comment by XhqGlorry11** [comment]
I AM TOTALLY CONFUSED WHY TENSORFLOW TEAM REJECT TO ADD GROUP CONVOLUTION FOR SUCH A LONG TIME! WHAT ARE YOU THINKING ABOUT!!!!!!

**Comment by ppwwyyxx** [comment]
I already added it in #25818 and this issue can be closed.

**Comment by chsigg** [comment]
Thank you very much Yuxin for adding grouped convolutions! It is a great feature to have, and we simply didn't have the cycles to do it ourselves.

I'm closing this issue.

**Comment by mha-py** [comment]
Would it be possible to upload the files that are changed by changing the two lines of code? That would be super helpful, not everyone wants to install visual 2017 and other stuff just for going around that argument checking in the function
